### PR TITLE
Allow for passing zero to disable buffering for methods that accept bufferSize and use FileStream

### PR DESCRIPTION
**Pull request by Marusyk**
@adamsitnik could you please check if I on the right way?

Issue #53497

### CODE REVIEW

**Review by adamsitnik — COMMENTED** [review]
@Marusyk overall it looks good to me, but some tests are failing. To run the tests locally:

```cmd
.\build.cmd -rc Release -subset clr+libs+libs.tests
.\.dotnet\dotnet.exe build -c Debug -f net6.0-windows .\src\libraries\System.IO.FileSystem\tests\System.IO.FileSystem.Tests.csproj /t:Test
```

If you want to run specific test you need to pass the following property:

```cmd
/p:XunitMethodName=Namespace.TypeName.MehtodName
```

File: src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ private static void ValidateHandle(SafeFileHandle handle, FileAccess access, int
             {
                 throw new ArgumentOutOfRangeException(nameof(access), SR.ArgumentOutOfRange_Enum);
             }
-            else if (bufferSize <= 0)
+            else if (bufferSize < 0)
             {
                 throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs:83]: The error message should also be changed.

File: src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs [hunk 2]
```diff
@@ -152,7 +152,7 @@ public StreamReader(Stream stream, Encoding? encoding = null, bool detectEncodin
             {
                 bufferSize = DefaultBufferSize;
             }
-            else if (bufferSize <= 0)
+            else if (bufferSize < 0)
             {
                 throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs:157]: Same for all of these.

**Comment by stephentoub** [comment]
@adamsitnik, if 0 is now to be considered valid, why do we still have tests like this?
https://github.com/dotnet/runtime/blob/8709d595f7105cbf73774ec762266bc56f99cf04/src/libraries/System.IO.FileSystem/tests/FileStream/ctor_sfh_fa_buffer.cs#L21-L30
Thanks.

**Comment by adamsitnik** [comment]
> @adamsitnik, if 0 is now to be considered valid, why do we still have tests like this?

I've changed that only for `FileStream` created from path. When I realized that the handle ctor is having the old check I've created #53497 to track this and this PR is about fixing that.

**Comment by stephentoub** [comment]
> I've created #53497 to track this and this PR is about fixing that.

Ok, but that issue mentions nothing about this public FileStream surface area.

**Comment by adamsitnik** [comment]
> Ok, but that issue mentions nothing about this public FileStream surface area.

Because I have a fix for that in #53669

**Comment by stephentoub** [comment]
> Because I have a fix for that in #53669

Ok, then this PR is blocked on that one.  The test failures here are because of that.

**Comment by adamsitnik** [comment]
In  #53669 we have fixed it for `FileStream` ctors that accept handles. In  https://github.com/dotnet/runtime/pull/53982#discussion_r650066127 we got to the conclusion that we are not going to change it for `StreamReader` and `StreamWriter`.

@Marusyk thank you for your help and apologies for inconvenience from our side

**Pull request closed without merging**